l10n_it_declaration_of_intent/tests/test_declaration_of_intent.py

```diff
@@ -0,0 +1,315 @@
+        post_used_amount = self.declaration1.used_amount
+        self.assertEqual(previous_used_amount, post_used_amount)
+        self.invoice_without_valid_taxes.action_post()
+        post_used_amount = self.declaration1.used_amount
+        self.assertEqual(previous_used_amount, post_used_amount)
+
+    def test_invoice_reopen_with_no_effect_on_declaration(self):
+        previous_used_amount = self.declaration1.used_amount
+        self.invoice1.action_post()
+        self.invoice1.button_cancel()
+        post_used_amount = self.declaration1.used_amount
+        self.assertEqual(previous_used_amount, post_used_amount)
+
+    def test_invoice_validation_under_declaration_limit(self):
+        previous_used_amount = self.declaration1.used_amount
+        self.invoice2.action_post()
+        post_used_amount = self.declaration1.used_amount
+        self.assertNotEqual(previous_used_amount, post_used_amount)
+
+    def test_invoice_validation_over_declaration_limit(self):
+        self.invoice2.action_post()
+        with self.assertRaises(UserError):
+            self.invoice3.action_post()
+
+    def test_invoice_reopen_with_effect_on_declaration(self):
+        previous_used_amount = self.declaration1.used_amount
+        self.invoice2.action_post()
+        self.invoice2.button_cancel()
+        post_used_amount = self.declaration1.used_amount
+        self.assertEqual(previous_used_amount, post_used_amount)
+
+    def test_refund(self):
+        self.invoice2.action_post()
+        previous_used_amount = self.declaration1.used_amount
+        self.refund1.action_post()
+        post_used_amount = self.declaration1.used_amount
+        self.assertNotEqual(previous_used_amount, post_used_amount)
+
+    def test_refund_with_amount_bigger_than_residual(self):
+        self.invoice2.action_post()
+        refund_form = Form(self.refund1)
+        with refund_form.invoice_line_ids.edit(0) as line_form:
+            line_form.quantity = 10
+        refund_form.save()
+
+        # Check that base amount has been updated
+        self.assertEqual(self.refund1.amount_untaxed, 1000)
+
+        # Refund goes over plafond: 100 + 1000 > 1000
+        self.assertEqual(self.declaration1.available_amount, 100)
+        self.assertEqual(self.refund1.amount_untaxed, 1000)
+        self.assertEqual(self.declaration1.limit_amount, 1000)
+        with self.assertRaises(UserError):
+            self.refund1.action_post()
+
+    def test_fiscal_position_no_declaration(self):
+        self.invoice4._onchange_date_invoice()
+        self.assertEqual(self.invoice4.fiscal_position_id.id, self.fiscal_position2.id)
+
+    def test_invoice_vendor_with_no_effect_on_declaration(self):
+        previous_used_amount = self.declaration4.used_amount
+        self.assertAlmostEqual(previous_used_amount, 0.0, 2)
+        self.invoice5.action_post()
+        post_used_amount = self.declaration4.used_amount
+        self.assertAlmostEqual(post_used_amount, 900.0, 2)
```

Comment:
Se ha senso considerare le varie currency_id, io direi di aggiungere test anche per quelle casistiche (fatture in valuta diversa dalla dichiarazione)

Reply:
Da risolvere in https://github.com/OCA/l10n-italy/issues/2428